File: src/utils/emotionAnalysis.ts
import { type BodyMarkings, type EmotionResult } from "@/types/bodyMap";

// Emotion detection patterns based on established body mapping research
const EMOTION_PATTERNS = {
  anger: {
    hot: ['head', 'chest', 'left-arm', 'right-arm', 'left-hand', 'right-hand'],
    warm: ['neck', 'abdomen'],
    indicators: ['Hot face and arms', 'Warm chest area', 'Activated upper body'],
    description: "Anger typically manifests as intense heat in the head, chest, and arms, representing the body's preparation for action."
  },
  happiness: {
    warm: ['head', 'chest', 'left-arm', 'right-arm', 'abdomen'],
    hot: ['chest'],
    indicators: ['Warm all over', 'Especially warm chest', 'Positive energy throughout'],
    description: "Happiness creates a warm, energized feeling throughout the body, with particular warmth in the chest and limbs."
  },
  sadness: {
    cold: ['left-arm', 'right-arm', 'left-leg', 'right-leg', 'left-hand', 'right-hand'],
    cool: ['chest', 'abdomen'],
    numb: ['chest'],
    indicators: ['Cold limbs', 'Reduced chest activity', 'Withdrawal of energy'],
    description: "Sadness shows as coldness in the limbs and reduced activity in the torso, reflecting emotional withdrawal."
  },
  fear: {
    hot: ['head', 'chest'],
    cold: ['left-arm', 'right-arm', 'left-leg', 'right-leg'],
    cool: ['abdomen'],
    indicators: ['Hot head and chest', 'Cold limbs', 'Fight-or-flight activation'],
    description: "Fear creates intense activation in the head and chest while causing coldness in the limbs."
  },
  anxiety: {
    hot: ['head', 'chest'],
    warm: ['abdomen'],
    cold: ['left-hand', 'right-hand', 'left-foot', 'right-foot'],
    indicators: ['Racing mind', 'Chest tightness', 'Cold extremities'],
    description: "Anxiety manifests as heat in the head and chest with coldness in hands and feet."
  },
  love: {
    warm: ['head', 'chest', 'left-arm', 'right-arm', 'abdomen'],
    hot: ['chest'],
    indicators: ['Warm everywhere', 'Glowing chest', 'Full-body activation'],
    description: "Love creates an all-encompassing warmth, especially strong in the chest area."
  },
  disgust: {
    cool: ['head', 'chest'],
    cold: ['abdomen'],
    numb: ['left-arm', 'right-arm'],
    indicators: ['Cool upper body', 'Cold stomach', 'Reduced limb activity'],
    description: "Disgust shows as coolness in the upper body and coldness in the stomach area."
  },
  surprise: {
    hot: ['head', 'chest'],
    warm: ['left-arm', 'right-arm'],
    indicators: ['Sudden head activation', 'Chest response', 'Alert limbs'],
    description: "Surprise creates sudden activation in the head and chest with alertness in the arms."
  },
  shame: {
    hot: ['head'],
    cold: ['chest', 'abdomen'],
    numb: ['left-arm', 'right-arm'],
    indicators: ['Hot face', 'Cold torso', 'Withdrawn limbs'],
    description: "Shame manifests as heat in the face while the body feels cold and withdrawn."
  },
  pride: {
    hot: ['head', 'chest'],
    warm: ['abdomen', 'left-arm', 'right-arm'],
    indicators: ['Expanded chest', 'Head held high', 'Strong posture'],
    description: "Pride creates warmth and expansion in the upper body, particularly the chest and head."
  }
};

export function analyzeEmotions(markings: BodyMarkings): EmotionResult[] {
  const results: EmotionResult[] = [];
  
  // Combine front and back markings
  const allMarkings = { ...markings.front, ...markings.back };
  
  // Count how many regions have sensations
  const markedRegions = Object.values(allMarkings).filter(sensation => sensation !== null).length;
  
  if (markedRegions === 0) {
    return [{
      emotion: "Neutral/Calm",
      confidence: 0.8,
      description: "No significant sensations detected, suggesting a calm or neutral emotional state.",
      patterns: ["No marked sensations"]
    }];
  }

  // Analyze each emotion pattern with improved sensitivity
  for (const [emotion, pattern] of Object.entries(EMOTION_PATTERNS)) {
    let matches = 0;
    let totalChecked = 0;
    const foundPatterns: string[] = [];

    // Check each sensation type in the pattern
    for (const [sensation, expectedRegions] of Object.entries(pattern)) {
      if (sensation === 'indicators' || sensation === 'description') continue;
      
      const regions = expectedRegions as string[];

      for (const region of regions) {
        totalChecked++;
        if (allMarkings[region as keyof typeof allMarkings] === sensation) {
          matches++;
          foundPatterns.push(`${sensation} ${region.replace('-', ' ')}`);
        }
      }
    }

    if (matches > 0 && totalChecked > 0) {
      // More generous confidence calculation
      const confidence = Math.min((matches / Math.max(totalChecked * 0.3, 1)), 1);
      
      // Lower threshold for inclusion
      if (confidence > 0.1) {
        results.push({
          emotion: emotion.charAt(0).toUpperCase() + emotion.slice(1),
          confidence,
          description: pattern.description,
          patterns: foundPatterns.length > 0 ? foundPatterns : pattern.indicators
        });
      }
    }
  }

  // If no strong matches, provide contextual suggestions based on individual sensations
  if (results.length === 0 || results.every(r => r.confidence < 0.2)) {
    const contextualResults = getContextualSuggestions(allMarkings);
    results.push(...contextualResults);
  }

  // Sort by confidence and return top results
  return results
    .sort((a, b) => b.confidence - a.confidence)
    .slice(0, 3);
}

// New function to provide contextual suggestions when no clear patterns match
function getContextualSuggestions(markings: Record<string, any>): EmotionResult[] {
  const suggestions: EmotionResult[] = [];
  
  // Count sensations by type
  const sensationCounts = {
    hot: 0, warm: 0, cool: 0, cold: 0, numb: 0
  };
  
  const sensationRegions: Record<string, string[]> = {
    hot: [], warm: [], cool: [], cold: [], numb: []
  };

  for (const [region, sensation] of Object.entries(markings)) {
    if (sensation && sensationCounts.hasOwnProperty(sensation)) {
      sensationCounts[sensation as keyof typeof sensationCounts]++;
      sensationRegions[sensation].push(region.replace('-', ' '));
    }
  }

  // Cold legs/limbs often indicates fear or anxiety
  if (sensationCounts.cold > 0) {
    const coldLimbs = sensationRegions.cold.filter(region => 
      region.includes('leg') || region.includes('foot') || region.includes('arm') || region.includes('hand')
    );
    
    if (coldLimbs.length > 0) {
      suggestions.push({
        emotion: "Fear/Anxiety",
        confidence: 0.6,
        description: "Cold sensations in the limbs often indicate fear or anxiety, as the body redirects blood flow to vital organs.",
        patterns: [`Cold in: ${coldLimbs.join(', ')}`]
      });
    }
  }

  // Hot head/face often indicates anger or stress
  if (sensationCounts.hot > 0) {
    const hotHead = sensationRegions.hot.filter(region => 
      region.includes('head') || region.includes('face')
    );
    
    if (hotHead.length > 0) {
      suggestions.push({
        emotion: "Anger/Stress",
        confidence: 0.5,
        description: "Heat in the head or face area typically indicates anger, frustration, or intense stress.",
        patterns: [`Hot in: ${hotHead.join(', ')}`]
      });
    }
  }

  // Numb areas suggest emotional shutdown or overwhelm
  if (sensationCounts.numb > 0) {
    suggestions.push({
      emotion: "Emotional Numbness",
      confidence: 0.5,
      description: "Numbness in body areas can indicate emotional shutdown, overwhelm, or disconnection.",
      patterns: [`Numb areas: ${sensationRegions.numb.join(', ')}`]
    });
  }

  // General cold suggests withdrawal or sadness
  if (sensationCounts.cold > sensationCounts.hot + sensationCounts.warm) {
    suggestions.push({
      emotion: "Sadness/Withdrawal",
      confidence: 0.4,
      description: "Predominant cold sensations often indicate sadness, withdrawal, or low energy states.",
      patterns: [`Cold dominance in: ${sensationRegions.cold.join(', ')}`]
    });
  }

  // General warmth suggests positive states
  if (sensationCounts.warm > sensationCounts.cold + sensationCounts.cool) {
    suggestions.push({
      emotion: "Contentment/Warmth",
      confidence: 0.4,
      description: "Predominant warm sensations often indicate contentment, love, or positive emotional states.",
      patterns: [`Warmth in: ${sensationRegions.warm.join(', ')}`]
    });
  }

  return suggestions;
}

File: backend/app/services/emotion_analysis.py
#!/usr/bin/env python3
"""
Emotion Analysis Service
Uses LLM service with priority: Gemini API -> OpenAI API -> Local patterns
"""

from typing import Dict, List
from .llm_service import LLMService

class EmotionAnalysisService:
    """Service for analyzing emotions from body sensations"""
    
    def __init__(self):
        # Initialize the LLM service with priority order
        self.llm_service = LLMService()
        
        # Keep the original patterns as a backup
        self.EMOTION_PATTERNS = {
            'hot': {
                'head': ['Anger', 'Stress', 'Frustration'],
                'chest': ['Anxiety', 'Excitement', 'Anger'],
                'stomach': ['Nervousness', 'Excitement', 'Anger'],
                'arms': ['Tension', 'Stress', 'Anger'],
                'legs': ['Tension', 'Stress', 'Anger']
            },
            'warm': {
                'head': ['Contentment', 'Happiness', 'Calm'],
                'chest': ['Love', 'Happiness', 'Contentment'],
                'stomach': ['Contentment', 'Happiness', 'Calm'],
                'arms': ['Relaxation', 'Contentment', 'Happiness'],
                'legs': ['Relaxation', 'Contentment', 'Happiness']
            },
            'cold': {
                'head': ['Fear', 'Shock', 'Sadness'],
                'chest': ['Fear', 'Sadness', 'Withdrawal'],
                'stomach': ['Fear', 'Anxiety', 'Sadness'],
                'arms': ['Fear', 'Withdrawal', 'Sadness'],
                'legs': ['Fear', 'Withdrawal', 'Sadness']
            },
            'cool': {
                'head': ['Calm', 'Peace', 'Relaxation'],
                'chest': ['Calm', 'Peace', 'Relaxation'],
                'stomach': ['Calm', 'Peace', 'Relaxation'],
                'arms': ['Calm', 'Peace', 'Relaxation'],
                'legs': ['Calm', 'Peace', 'Relaxation']
            },
            'numb': {
                'head': ['Shock', 'Disconnection', 'Trauma'],
                'chest': ['Shock', 'Disconnection', 'Trauma'],
                'stomach': ['Shock', 'Disconnection', 'Trauma'],
                'arms': ['Shock', 'Disconnection', 'Trauma'],
                'legs': ['Shock', 'Disconnection', 'Trauma']
            }
        }
    
    def analyze_emotions(self, body_markings: Dict[str, str], view: str) -> Dict:
        """Analyze emotions using the prioritized LLM service"""
        try:
            print(f"🧠 Analyzing emotions with LLM service...")
            print(f"   Body markings: {body_markings}")
            print(f"   View: {view}")
            
            # Use the LLM service (Gemini -> OpenAI -> Local patterns)
            result = self.llm_service.analyze_emotions(body_markings, view)
            
            if result:
                print(f"✅ Emotion analysis successful: {result.get('emotion', 'Unknown')}")
                return result
            else:
                print("⚠️  LLM service failed, using local patterns")
                return self._local_pattern_analysis(body_markings)
                
        except Exception as e:
            print(f"❌ Emotion analysis failed: {e}")
            print("🔄 Falling back to local pattern analysis")
            return self._local_pattern_analysis(body_markings)
    
    def _local_pattern_analysis(self, body_markings: Dict[str, str]) -> Dict:
        """Fallback to local pattern matching"""
        # Count different sensation types
        hot_count = sum(1 for sensation in body_markings.values() if sensation == 'hot')
        warm_count = sum(1 for sensation in body_markings.values() if sensation == 'warm')
        cold_count = sum(1 for sensation in body_markings.values() if sensation == 'cold')
        cool_count = sum(1 for sensation in body_markings.values() if sensation == 'cool')
        numb_count = sum(1 for sensation in body_markings.values() if sensation == 'numb')
        
        # Analyze patterns
        results = []
        
        # Pattern 1: High energy (hot sensations)
        if hot_count >= 2:
            results.append({
                'emotion': 'Anger/Excitement',
                'confidence': 0.8,
                'description': 'Multiple hot sensations suggest high energy emotions like anger or excitement',
                'patterns': [f'{hot_count} hot sensations'],
                'source': 'local_pattern_analysis'
            })
        
        # Pattern 2: Positive warmth
        if warm_count >= 2 and hot_count == 0:
            results.append({
                'emotion': 'Happiness/Contentment',
                'confidence': 0.7,
                'description': 'Warm sensations without hot suggest positive emotional state',
                'patterns': [f'{warm_count} warm sensations'],
                'source': 'local_pattern_analysis'
            })
        
        # Pattern 3: Withdrawal (cold sensations)
        if cold_count >= 2:
            results.append({
                'emotion': 'Sadness/Withdrawal',
                'confidence': 0.7,
                'description': 'Cold sensations suggest emotional withdrawal or sadness',
                'patterns': [f'{cold_count} cold sensations'],
                'source': 'local_pattern_analysis'
            })
        
        # Pattern 4: Anxiety (mixed hot and cold)
        if hot_count >= 1 and cold_count >= 1:
            results.append({
                'emotion': 'Anxiety/Stress',
                'confidence': 0.6,
                'description': 'Mixed hot and cold sensations suggest anxiety or stress',
                'patterns': [f'{hot_count} hot, {cold_count} cold'],
                'source': 'local_pattern_analysis'
            })
        
        # Pattern 5: Numbness
        if numb_count >= 1:
            results.append({
                'emotion': 'Disconnection/Shock',
                'confidence': 0.6,
                'description': 'Numb sensations suggest emotional disconnection or shock',
                'patterns': [f'{numb_count} numb sensations'],
                'source': 'local_pattern_analysis'
            })
        
        # If no clear patterns, provide general analysis
        if not results:
            total_sensations = len([s for s in body_markings.values() if s])
            if total_sensations == 0:
                results.append({
                    'emotion': 'Neutral/Calm',
                    'confidence': 0.8,
                    'description': 'No significant sensations detected, suggesting a calm state',
                    'patterns': ['no sensations'],
                    'source': 'local_pattern_analysis'
                })
            else:
                results.append({
                    'emotion': 'Mixed/Complex',
                    'confidence': 0.5,
                    'description': f'Complex pattern of {total_sensations} sensations suggests mixed emotions',
                    'patterns': [f'{total_sensations} total sensations'],
                    'source': 'local_pattern_analysis'
                })
        
        return results
    
    def get_service_status(self) -> Dict:
        """Get the status of the emotion analysis service"""
        return {
            'service': 'emotion_analysis',
            'status': 'active',
            'llm_providers': len(self.llm_service.providers),
            'primary_provider': 'Gemini API' if self.llm_service.providers else 'Local patterns only',
            'fallback_available': True
        }
